packages/vite/src/node/server/shortcuts.ts

```diff
@@ -0,0 +1,95 @@
+import colors from 'picocolors'
+import type { ViteDevServer } from '..'
+import { openBrowser } from './openBrowser'
+
+export function bindShortcuts(server: ViteDevServer): void {
+  if (!server.httpServer) return
+
+  server.config.logger.info(
+    `  > Shortcuts: ` +
+      SHORTCUTS.map((shortcut) => {
+        return `"${shortcut.key}" ${shortcut.name}`
+      }).join(', ')
+  )
+
+  let actionRunning = false
+
+  const keypressHandler = async (input: string) => {
+    // ctrl+c or ctrl+d
+    if (input === '\x03' || input === '\x04') {
+      return process.kill(process.pid, 'SIGINT')
+    }
+
+    const shortcut = SHORTCUTS.find((shortcut) => shortcut.key === input)
+
+    if (!shortcut) {
+      return
+    }
+
+    if (actionRunning) {
+      return
+    }
+
+    actionRunning = true
+    await shortcut.action(server)
+    actionRunning = false
+  }
+
+  if (process.stdin.isTTY) {
+    process.stdin.setRawMode(true)
+  }
+
+  process.stdin.on('keypress', keypressHandler).setEncoding('utf8').resume()
```

Comment:
The `keypress` event requires [`readline.emitKeypressEvents(process.stdin)`](https://nodejs.org/api/readline.html#readlineemitkeypresseventsstream-interface) beforehand.

I think the reason Vitest uses this instead of just the `data` event is because its shortcuts include a prompt, where it needs to pause keypress handling to properly receive the input. Since our shortcuts don't have any prompts, we probably don't need this. Sorry for having suggested this change in my original review.